File: app/src/main/java/com/example/rcapp/CameraActivity.java
package com.example.rcapp;

import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.media.Image;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.core.Camera;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageCapture;
import androidx.camera.core.ImageCaptureException;
import androidx.camera.core.ImageProxy;
import androidx.camera.core.Preview;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.camera.view.PreviewView;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.example.rcapp.databinding.ActivityCameraBinding;
import com.google.common.util.concurrent.ListenableFuture;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.net.URI;
import java.nio.ByteBuffer;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import android.Manifest;

import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okio.ByteString;

public class CameraActivity extends AppCompatActivity {
    private ActivityCameraBinding binding;
    private static final String TAG = "CameraActivity";
    private static final String WEBSOCKET_URL = "ws://192.168.3.7:8766"; // WebSocket 服务器地址
    private PreviewView previewView;
    private ImageAnalysis imageAnalysis;
    private ProcessCameraProvider cameraProvider;
    private ExecutorService cameraExecutor;
    private WebSocket webSocket;
    private static final int REQUEST_CODE_PERMISSIONS = 10;
    private static final String[] REQUIRED_PERMISSIONS = new String[]{
            Manifest.permission.CAMERA
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding=ActivityCameraBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());previewView = findViewById(R.id.viewFinder);
        //启动WebSocket，与Python后端连接，传输视频流
        setupWebSocket();
        if (allPermissionsGranted()) {
            //权限足够则开启相机
            startCamera();
        } else {
            //否则请求权限
            requestPermissions();
        }
        cameraExecutor = Executors.newSingleThreadExecutor();
    }

    private void startCamera() {
        //加载相机，进行后续监听回调
        ListenableFuture<ProcessCameraProvider> cameraProviderFuture =
                ProcessCameraProvider.getInstance(this);

        cameraProviderFuture.addListener(() -> {
            try {
                //获取相机实例
                cameraProvider = cameraProviderFuture.get();
                // 预览用例
                Preview preview = new Preview.Builder().build();
                preview.setSurfaceProvider(previewView.getSurfaceProvider());
//                // 图像分析用例
                imageAnalysis = new ImageAnalysis.Builder()
                        .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
                        .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                        .build();

                imageAnalysis.setAnalyzer(cameraExecutor, new ImageAnalysis.Analyzer() {
                    @Override
                    @androidx.camera.core.ExperimentalGetImage
                    public void analyze(@NonNull ImageProxy imageProxy) {
                        Image image = imageProxy.getImage();
                        if (image != null) {
                            processImage(image);
                        }
                        imageProxy.close();
                    }
                });
//                // 选择后置摄像头
//                CameraSelector cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA;
//                // 解绑所有用例
//                cameraProvider.unbindAll();
//                // 绑定用例到相机
//                Camera camera = cameraProvider.bindToLifecycle(
//                        this,
//                        cameraSelector,
//                        preview,
//                        imageAnalysis
//                );
                // 创建 ImageCapture 用例
                ImageCapture imageCapture = new ImageCapture.Builder().build();

                // 解绑所有用例
                cameraProvider.unbindAll();
                // 选择后置摄像头
                CameraSelector cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA;
                // 绑定用例到相机
                Camera camera = cameraProvider.bindToLifecycle(
                        this,
                        cameraSelector,
                        preview,
                        imageAnalysis,
                        imageCapture // 绑定 ImageCapture
                );

                // 设置拍照按钮的点击事件
                findViewById(R.id.camera_photo_btn).setOnClickListener(v -> {
                    // 创建文件以保存图片
                    File photoFile = new File(getExternalFilesDir(null), "photo.jpg");
                    ImageCapture.OutputFileOptions outputFileOptions =
                            new ImageCapture.OutputFileOptions.Builder(photoFile).build();

                    // 拍照
                    imageCapture.takePicture(outputFileOptions, ContextCompat.getMainExecutor(this),
                            new ImageCapture.OnImageSavedCallback() {
                                @Override
                                public void onImageSaved(@NonNull ImageCapture.OutputFileResults outputFileResults) {
                                    Toast.makeText(CameraActivity.this, "拍照成功: " + photoFile.getAbsolutePath(), Toast.LENGTH_SHORT).show();
                                }

                                @Override
                                public void onError(@NonNull ImageCaptureException exception) {
                                    Log.e(TAG, "拍照失败: ", exception);
                                }
                            });
                });


            } catch (ExecutionException | InterruptedException e) {
                Log.e(TAG, "Error starting camera: ", e);
            }
        }, ContextCompat.getMainExecutor(this));
    }

    private void processImage(Image image) {
        Image.Plane[] planes = image.getPlanes();
        ByteBuffer buffer = planes[0].getBuffer();
        int pixelStride = planes[0].getPixelStride();
        int rowStride = planes[0].getRowStride();
        int rowPadding = rowStride - pixelStride * image.getWidth();

        // 创建位图
        Bitmap bitmap = Bitmap.createBitmap(
                image.getWidth() + rowPadding / pixelStride,
                image.getHeight(),
                Bitmap.Config.ARGB_8888
        );
        bitmap.copyPixelsFromBuffer(buffer);
        sendImageToServer(bitmap);

        runOnUiThread(() -> {
            // 这里可以对位图进行处理或显示
            // 例如: imageView.setImageBitmap(bitmap);
        });

        // 可以选择保存图像
        // saveImage(bitmap);
    }
    private void sendImageToServer(Bitmap bitmap) {
        // 将位图转换为字节数组并发送到 WebSocket
        // 这里可以将 Bitmap 转换为 JPEG 格式并发送
        ByteArrayOutputStream stream = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.JPEG, 90, stream);
        byte[] byteArray = stream.toByteArray();
        webSocket.send(ByteString.of(byteArray)); // 发送字节数组
    }
    private void saveImage(Bitmap bitmap) {
        ContentValues values = new ContentValues();
        values.put(MediaStore.Images.Media.DISPLAY_NAME,
                "Image_" + System.currentTimeMillis() + ".jpg");
        values.put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg");

        values.put(MediaStore.Images.Media.RELATIVE_PATH, "Pictures/CameraX");
        values.put(MediaStore.Images.Media.IS_PENDING, 1);

        ContentResolver resolver = getContentResolver();
        Uri collection = MediaStore.Images.Media.getContentUri(
                MediaStore.VOLUME_EXTERNAL_PRIMARY);
        Uri imageUri = resolver.insert(collection, values);

        try {
            if (imageUri != null) {
                try (OutputStream out = resolver.openOutputStream(imageUri)) {
                    if (out != null) {
                        bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out);
                    }
                }

                values.clear();
                values.put(MediaStore.Images.Media.IS_PENDING, 0);
                resolver.update(imageUri, values, null, null);
            }
        } catch (IOException e) {
            Log.e(TAG, "Failed to save image", e);
        }
    }

    private boolean allPermissionsGranted() {
        for (String permission : REQUIRED_PERMISSIONS) {
            if (ContextCompat.checkSelfPermission(this, permission)
                    != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    private void requestPermissions() {
        ActivityCompat.requestPermissions(
                this,
                REQUIRED_PERMISSIONS,
                REQUEST_CODE_PERMISSIONS
        );
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera();
            } else {
                Toast.makeText(this,
                        "Permissions not granted by the user.",
                        Toast.LENGTH_SHORT).show();
                finish();
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        cameraExecutor.shutdown();
    }
    private void setupWebSocket() {
        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder().url(WEBSOCKET_URL).build();

        webSocket = client.newWebSocket(request, new WebSocketListener() {
            @Override
            public void onOpen(@NonNull WebSocket webSocket, @NonNull Response response) {
                Log.d(TAG, "WebSocket connection opened");
            }

            @Override
            public void onMessage(@NonNull WebSocket webSocket, @NonNull String text) {
                Log.d(TAG, "Message from server: " + text);
            }

            @Override
            public void onMessage(@NonNull WebSocket webSocket, @NonNull ByteString bytes) {
                Log.d(TAG, "Message from server: " + bytes.hex());
            }

            @Override
            public void onClosed(@NonNull WebSocket webSocket, int code, @NonNull String reason) {
                Log.d(TAG, "WebSocket connection closed: " + reason);
            }

            @Override
            public void onFailure(@NonNull WebSocket webSocket, @NonNull Throwable t, Response response) {
                Log.e(TAG, "WebSocket error: ", t);
            }
        });
    }

}

File: app/src/main/java/com/example/rcapp/activity/CameraActivity.kt
package com.example.rcapp.activity

import android.Manifest.permission
import android.content.ContentValues
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.media.Image
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import android.widget.Toast
import androidx.annotation.OptIn
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.CameraSelector
import androidx.camera.core.ExperimentalGetImage
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageCapture
import androidx.camera.core.ImageCapture.OutputFileOptions
import androidx.camera.core.ImageCapture.OutputFileResults
import androidx.camera.core.ImageCaptureException
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.example.rcapp.R
import com.example.rcapp.databinding.ActivityCameraBinding
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okhttp3.Request.Builder
import okio.ByteString
import java.io.ByteArrayOutputStream
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.UUID
import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class CameraActivity  : BleServiceBaseActivity() {
    private var binding: ActivityCameraBinding? = null
    private var previewView: PreviewView? = null
    private var cameraProvider: ProcessCameraProvider? = null
    private var cameraExecutor: ExecutorService? = null
    private var webSocket: WebSocket? = null
    private var context: Context? = null

    override fun onBluetoothServiceConnected() {
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCameraBinding.inflate(layoutInflater)
        setContentView(binding!!.root)

        previewView = findViewById(R.id.viewFinder)
        //先设置网络通信接口
        setupWebSocket()

        if (allPermissionsGranted()) {
            //权限足够则开启相机
            startCamera()
        } else {
            //否则请求权限
            requestPermissions()
        }
        //创建一个相机线程
        cameraExecutor = Executors.newSingleThreadExecutor()
    }

    //开启相机
    private fun startCamera() {
        /**
         * ListenableFuture 是一个接口，来自 Guava 库，允许异步操作的结果在完成后被监听
         * ProcessCameraProvider是 CameraX 提供的一个类，用于管理相机生命周期和绑定相机用例。
         * context作为参数，是ProcessCameraProvider提供者
         */
        val cameraProviderFuture = ProcessCameraProvider.getInstance(this)
        // 获取 cameraProviderFuture 的异步结果，即ProcessCameraProvider.getInstance(this)的获取结果
        cameraProviderFuture.addListener({
            try {
                // 获取 ProcessCameraProvider 实例，即cameraProvider
                cameraProvider = cameraProviderFuture.get()
                // 绑定相机用例（如预览、拍照）
                bindCameraUseCases()
            } catch (e: ExecutionException) {
                Log.e(TAG, "Error starting camera: ", e)
            } catch (e: InterruptedException) {
                Log.e(TAG, "Error starting camera: ", e)
            }
        }, ContextCompat.getMainExecutor(this)) // 在主线程上执行监听器
    }

    @OptIn(ExperimentalGetImage::class)//实验性API声明
    private fun bindCameraUseCases() {
        //PreviewView 是 UI 中用于显示相机预览的视图
        val preview = Preview.Builder().build()
        preview.setSurfaceProvider(previewView!!.surfaceProvider)

        //创建一个图像分析用例，用于处理实时相机帧
        val imageAnalysis = ImageAnalysis.Builder()
            .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
            .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
            .build()
        //设置分析结果处理
        imageAnalysis.setAnalyzer(
            cameraExecutor!!
        ) { imageProxy ->
            //ImageProxy 是图像帧
            //图像帧转化为Image对象
            val image = imageProxy.image
            if (image != null) {
                processImage(image)
            }
            imageProxy.close()
        }
        //创建一个图像捕获用例，用于拍摄照片
        val imageCapture = ImageCapture.Builder().build()
        //解绑先前相机的所有用例
        cameraProvider!!.unbindAll()
        //CameraSelector.DEFAULT_BACK_CAMERA表示选择后置相机
        val cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA
        //绑定各个用例到相机生命周期
        cameraProvider!!.bindToLifecycle(this, cameraSelector, preview, imageAnalysis, imageCapture)
        //拍照事件监听
        binding?.cameraPhotoBtn?.setOnClickListener {
            takePhoto(imageCapture)
        }
    }

    /**
     * 拍照功能具体实现
     */
    private fun takePhoto(imageCapture: ImageCapture) {
        // 使用当前时间创建一个唯一的文件名
        val name = SimpleDateFormat("yyyyMMdd_msys", Locale.US).format(System.currentTimeMillis())
        // 创建 ContentValues 来存储图片的元数据
        val contentValues = ContentValues().apply {
            // 设置图片的显示名称
            put(MediaStore.Images.Media.DISPLAY_NAME, "$name.jpg")
            // 设置图片的 MIME 类型
            put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg")
            // 指定图片存储的相对路径（相册目录）
            put(MediaStore.Images.Media.RELATIVE_PATH, "Pictures/MyApp")
        }

        // 创建 OutputFileOptions，指定内容解析器和存储位置
        val outputOptions = OutputFileOptions.Builder(
            contentResolver,
            MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
            contentValues
        ).build()

        // 调用 takePicture 方法拍摄照片
        imageCapture.takePicture(outputOptions, ContextCompat.getMainExecutor(this),
            object : ImageCapture.OnImageSavedCallback {
                // 当图片成功保存时调用
                override fun onImageSaved(outputFileResults: OutputFileResults) {
                    Toast.makeText(
                        this@CameraActivity,
                        "拍照成功",
                        Toast.LENGTH_SHORT
                    ).show()
                }

                // 当拍摄过程中发生错误时调用
                override fun onError(exception: ImageCaptureException) {
                    Log.e(TAG, "拍照失败: ", exception) // 记录错误日志
                }
            })
    }

    private fun processImage(image: Image) {
        // 获取图像平面的缓冲区
        val buffer = image.planes[0].buffer
        // 创建一个位图，指定宽度、高度和颜色配置
        val bitmap = Bitmap.createBitmap(image.width, image.height, Bitmap.Config.ARGB_8888)
        // 将缓冲区中的像素数据复制到位图中
        bitmap.copyPixelsFromBuffer(buffer)
        // 将位图发送到服务器进行处理
        sendImageToServer(bitmap)
    }

    private fun sendImageToServer(bitmap: Bitmap) {
        // 创建一个字节数组输出流来存储压缩后的图像数据
        val stream = ByteArrayOutputStream()
        // 将位图压缩为JPEG格式，质量为90%
        bitmap.compress(Bitmap.CompressFormat.JPEG, 90, stream)
        // 将输出流转换为字节数组
        val byteArray = stream.toByteArray()
        // 将字节数组转换为 ByteString，以便通过 WebSocket 发送
        val byteString = ByteString.of(*byteArray)
        // 通过 WebSocket 发送字节数据
        webSocket?.send(byteString)
    }

    //请求摄像机相关权限
    private fun allPermissionsGranted(): Boolean {
        for (permission in REQUIRED_PERMISSIONS) {
            if (ContextCompat.checkSelfPermission(
                    this,
                    permission
                ) != PackageManager.PERMISSION_GRANTED
            ) {
                return false
            }
        }
        return true
    }

    /**
     * 相机权限请求
     */
    private fun requestPermissions() {
        ActivityCompat.requestPermissions(this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS)
    }

    /**
     * 相机权限请求回调结果
     */
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera()
            } else {
                Toast.makeText(this, "Permissions not granted by the user.", Toast.LENGTH_SHORT)
                    .show()
                finish()
            }
        }
    }

    /**
     * 关闭相机线程
     */
    override fun onDestroy() {
        super.onDestroy()
        cameraExecutor!!.shutdown()
    }

    //WebSocket设置，与后端通信，传输视频流
    private fun setupWebSocket() {
        val client = OkHttpClient()
        val request: Request = Builder().url(WEBSOCKET_URL).build()
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "WebSocket connection opened")
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                Log.d(TAG, "Message from server: $text")
            }

            //16进制数据接收
            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                Log.d(TAG, "Message from server: " + bytes.hex())
                // 调用 Service 方法进行数据传输
                if (bluetoothService != null) {
                    val characteristicUUID = UUID.fromString("0000ffe1-0000-1000-8000-00805f9b34fb")
                    bluetoothService!!.writeData(characteristicUUID, bytes.toByteArray())
                }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(
                    TAG,
                    "WebSocket connection closed: $reason"
                )
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "WebSocket error: ", t)
            }
        })
    }
    companion object {
        private const val TAG = "CameraActivity"
        //后端地址端口
        private const val WEBSOCKET_URL = "ws://192.168.31.180:8766"
        //请求编号
        private const val REQUEST_CODE_PERMISSIONS = 10
        //要请求的相机权限
        private val REQUIRED_PERMISSIONS = arrayOf(permission.CAMERA)
    }
}